frontend/src/app/services/auth.service.ts

```diff
@@ -0,0 +1,153 @@
+import { Injectable } from '@angular/core';
+import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
+import { SessionService } from './session.service';
+import { Router } from '@angular/router';
+import { BehaviorSubject, Observable, catchError, map, mergeMap, of, tap } from 'rxjs';
+import { User, UserResponse } from '../user/models/user';
+import { encodeUserData, parseUserData } from '../user/utils';
+import _ from 'underscore';
+import { HttpClient } from '@angular/common/http';
+
+@Injectable({
+    providedIn: 'root'
+})
+export class AuthService {
+    private currentUserSubject$ = new BehaviorSubject<User | null | undefined>(undefined);
+
+    currentUser$: Observable<User | null | undefined> = this.currentUserSubject$.pipe();
+    isAuthenticated$: Observable<boolean | undefined> = this.currentUserSubject$.pipe(
+        map(_.isObject)
+    )
+
+    private authRoute(route: string): string {
+        return `/users/${route}`
+    }
+
+    constructor(
+        private sessionService: SessionService,
+        private router: Router,
+        private http: HttpClient,
+    ) {
+        this.sessionService.expired.pipe(
+            takeUntilDestroyed()
+        ).subscribe(() => this.logout());
+        this.setInitialAuth();
+    }
+
+    private setAuth(user: User | null): void {
+        this.currentUserSubject$.next(user);
+    }
+
+    private purgeAuth(): void {
+        this.currentUserSubject$.next(null);
+    }
+
+    private checkUser(): Observable<User | null> {
+        return this.http.get<UserResponse>(this.authRoute('user/')).pipe(
+            catchError(error => of(null)),
+            map(parseUserData),
+        );
+    }
+
+    private setInitialAuth(): void {
+        this.checkUser()
+            .pipe(takeUntilDestroyed())
+            .subscribe(user =>
+                this.setAuth(user),
+            );
+    }
+
+    public currentUser(): User | null | undefined {
+        return this.currentUserSubject$.value;
+    }
+
+    /**
+     * Logs in, retrieves user response, transforms to User object
+     */
+    public login(username: string, password: string): Observable<User | null> {
+        return this.http.post<{ key: string }>(
+            this.authRoute('login/'),
+            { username, password }
+        ).pipe(
+            mergeMap(() => this.checkUser()),
+            tap(data => this.setAuth(data)),
+        );
+    }
+
+    public logout(redirectToLogin: boolean = false): void {
+        this.purgeAuth();
+        this.http.post(
+            this.authRoute('logout/'),
+            {}
+        ).subscribe((data) => {
+            console.log(data);
```

Comment:
Residual console.log().

(I'll be working on this service in my PR anyway, so you may also ignore it for yours.)